### ISSUE TITLE
Ability to export routines that accept / return `string`s

### ISSUE THREAD

**Issue by bradcray** [chapel-lang/chapel#12892]
As a Chapel programmer, I'd like to be able to pass strings into exported Chapel routines, operating on them within Chapel as type `string`, and to return `string` expressions from exported routines.  In doing so, I'd like the strings to be represented in the external language using the closest natural string type in that language (that is, I would like the Chapel compiler to emit whatever stub code is required to wrap/unwrap our string representation into the one used by the external language; similar to how we convert between Chapel's default arrays and Fortran's arrays in `--library-fortran` mode).  I'd like this behavior and capability because I consider strings to be a fundamental and core type, and in order to make working with strings as natural and hassle-free as possible in both languages.

As an example, given the Chapel routine:

```chapel
export proc appendYo(in arg: string): string {
  return arg + ", yo";
}
```

I'd want a `--library[-c]` compilation to expose the following prototype for use from C:

```c
[const?] char* appendYo([const?] char* arg);
```

I imagine that the Chapel compiler would generate a stub version of `appendYo()` that would wrap the incoming `char*` into a Chapel `string`; pass it to the (pure / normal) Chapel `appendYo_chpl()` routine which would operate on it, returning a new Chapel string; the stub code would then unwrap this Chapel string back into a `char*` returning it to C.  The result is that I should be able to make a call like the following from C and have it work:

```c
printf("My name is %s", appendYo("Brad"));
```

while also being able to still call the exported routine from Chapel (which would call directly into `appendYo_chpl()`, bypassing the stub routine):

```chapel
writeln("My name is ", appendYo("Brad"));
```

Similarly, a `--library-python` compilation would provide a Python module with an `appendYo()` routine that can be called from Python, passing and returning Python's unicode strings (presumably using Cython's `str` type under the covers to do so and converting between it and Chapel's string type in the stub code), permitting me make a call like the following from Python (with the standard caveat that I'm terrible at Python, though I remembered not to type a semicolon!):

```python
print ("My name is %s" % (appendYo("Brad")))
```

(while still being able to also call it from Chapel of course).

**Comment by bradcray** [comment]
Relating this issue to #10636, I think the proposal here is closest to approach A in that issue's description, though using `[const?] char*` instead of `c_string` on the C side of things.

Lydia wrapped up that issue with "We do want to eventually export the string type itself."  Here, I'm proposing the ability to export routines that accept or return strings without requiring the external language programmer to interact with Chapel's string type directly.  For example, I think it'd be a loss if the external user had to deal with the Chapel string type by doing something like `printf("%s", chpl_string_to_char_star(appendYo(char_star_to_chpl_string("Brad")));` or otherwise converting between the language's native string type and Chapel's.  I think string types are sufficiently primitive and core in most languages that we should provide direct support for passing strings back and forth without tying the effort to the more general effort of coming up with a way to export user-defined Chapel types for use in external languages (which was where some of the concluding notes on #10636 seemed to be heading).  As alluded to above, I think the current Fortran array passing is a good model for making a non-trivial type work naturally in both languages and is what we should try to emulate for strings as well.

(I think we may also ultimately want a way to pass values of Chapel's string type back and forth between exported routines without doing this unwrapping, but I don't think it should be the common case).

**Comment by bradcray** [comment]
@lydia-duncan, @dlongnecke-cray: I think we've got this now, is that right?  Can this be closed?

**Comment by bradcray** [comment]
(and if so, want to cite the relevant PRs and close it?)

**Comment by lydia-duncan** [comment]
Resolved by #13357 and #13732 (was waiting to make sure David had a chance to object if he knew anything I didn't)
